cmd/cloud_sql_proxy/proxy.go

```diff
@@ -221,62 +221,77 @@ var validNets = func() map[string]bool {
+	}
+	// Parse the instance connection name - everything before the "=".
+	ret.Instance = args[0]
+	proj, _, name := util.SplitName(ret.Instance)
+	if proj == "" || name == "" {
+		return instanceConfig{}, fmt.Errorf("invalid instance connection string: must be in the form `project:region:instance-name`; invalid name was %q", args[0])
+	}
+	// Parse the instance options - everything before the "=".
+	eM := "invalid instance options: must be in the form `unix:/path/to/socket`, `tcp:port`, `tcp:host:port`; invalid option was %q"
+	if len(args) == 2 {
+		opts := strings.Split(args[1], ":")
+		if opts[0] == "unix" {
+			// Listen via specified unix socket.
+			if len(opts) != 2 {
+				return instanceConfig{}, fmt.Errorf(eM, args[1])
+			}
+			ret.Network = "unix"
+			if strings.HasPrefix(opts[1], "/") {
+				ret.Address = opts[1] // Root path.
+			} else {
+				ret.Address = filepath.Join(dir, opts[1])
+			}
+		} else {
+			// Listen via the specified dial address.
+			ret.Network = opts[0]
+			switch len(opts) {
+			default:
+				return instanceConfig{}, fmt.Errorf(eM, args[1])
+			case 2:
+				// No "host" part of the address. Be safe and assume that they want a loopback address.
+				addr, ok := loopbackForNet[opts[0]]
+				if !ok {
+					return ret, fmt.Errorf("invalid %q: unrecognized network %v", args[1], opts[0])
+				}
+				ret.Address = fmt.Sprintf("%s:%s", addr, opts[1])
+			case 3:
+				// User provided a host and port; use that.
+				ret.Address = fmt.Sprintf("%s:%s", opts[1], opts[2])
 			}
-			ret.Address = fmt.Sprintf("%s:%s", addr, spl[1])
-		case 3:
-			// User provided a host and port; use that.
-			ret.Network = spl[0]
-			ret.Address = fmt.Sprintf("%s:%s", spl[1], spl[2])
 		}
 	} else {
-		sql, err := sqladmin.New(cl)
-		if err != nil {
-			return instanceConfig{}, err
-		}
-		sql.BasePath = *host
-		ret.Instance = instance
-		// Default to unix socket.
+		// Default to listening via unix socket in specified directory
 		ret.Network = "unix"
+		ret.Address = filepath.Join(dir, instance)
+	}
 
-		proj, _, name := util.SplitName(instance)
-		if proj == "" || name == "" {
-			return instanceConfig{}, fmt.Errorf("invalid instance name: must be in the form `project:region:instance-name`; invalid name was %q", instance)
-		}
-		// We allow people to omit the region due to historical reasons. It'll
-		// fail later in the code if this isn't allowed, so just assume it's
-		// allowed until we actually need the region in this API call.
-		in, err := sql.Instances.Get(proj, name).Do()
-		if err != nil {
+	// Use the SQL Admin API to verify compatibility with the instance.
+	sql, err := sqladmin.New(cl)
+	if err != nil {
+		return instanceConfig{}, err
+	}
+	sql.BasePath = *host
+	inst, err := sql.Instances.Get(proj, name).Do()
+	if err != nil {
+		return instanceConfig{}, err
+	}
+	if inst.BackendType == "FIRST_GEN" {
+		logging.Errorf("WARNING: proxy client does not support first generation Cloud SQL instances.")
```

Comment:
doesn't look like a warning. but i notice there's the word "WARNING" everywhere even though things are not warnings but actual fatal errors.

Reply:
leaving: existing behavior